### PR TITLE
Service Discovery: Add ECS Params schema

### PR DESCRIPTION
**Pull request by PettitWesley**
By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: ecs-cli/modules/utils/compose/ecs_params_reader.go [hunk 1]
```diff
@@ -104,6 +105,49 @@ type AwsVpcConfiguration struct {
...
+	ID          string `yaml:"id"`
+	Name        string `yaml:"name"`
+	Description string `yaml:"description"`
+}
+
+// PublicDNSNamespace holds information related to Route53 public DNS namespaces
+type PublicDNSNamespace struct {
+	ID   string `yaml:"id"`
+	Name string `yaml:"name"`
+}
```
Comment (SoManyHs) [inline ecs-cli/modules/utils/compose/ecs_params_reader.go]: (No-op musings on encapsulation): Not sure how go-yaml handles embedding, but I wonder if it's possible to use composition for the DNSNamespace types since there's some repetition...

File: ecs-cli/modules/utils/compose/ecs_params_reader.go [hunk 2]
```diff
@@ -104,6 +105,49 @@ type AwsVpcConfiguration struct {
...
+// DNSConfig holds the dns configuration for Service Discovery Services
+type DNSConfig struct {
+	Type string `yaml:"type"`
+	TTL  *int64 `yaml:"ttl"`
+}
+
+// HealthCheckCustomConfig
+type HealthCheckCustomConfig struct {
+	FailureThreshold *int64 `yaml:"failure_threshold"`
+}
```
Comment (SoManyHs) [inline ecs-cli/modules/utils/compose/ecs_params_reader.go]: Does `HealthCheckCustomConfig` contain other fields we plan on supporting in the future?

File: ecs-cli/modules/utils/compose/ecs_params_reader.go [hunk 3]
```diff
@@ -104,6 +105,49 @@ type AwsVpcConfiguration struct {
 
 type AssignPublicIp string
 
+// ServiceDiscovery hold information related to ECS/Route53 Service Discovery
+type ServiceDiscovery struct {
```
[the remark this answers is not in the document]
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader.go]: It doesn't, because the conclusion from that discussion was that its ok to not support that field (it can be added later if customers request it).

File: ecs-cli/modules/utils/compose/ecs_params_reader_test.go [hunk 4]
```diff
@@ -512,6 +512,65 @@ task_definition:
...
+		serviceDiscovery := ecsParams.RunParams.ServiceDiscovery
+		assert.Equal(t, "nginx", serviceDiscovery.ContainerName, "Expected ContainerName to match")
+		assert.Equal(t, aws.Int64(80), serviceDiscovery.ContainerPort, "Expected ContainerPort to match")
+		assert.Equal(t, "vpc-8BAADF00D", serviceDiscovery.PrivateDNSNamespace.VPC, "Expected VPC to match")
+		assert.Equal(t, "ns-CA15CA15CA15CA15", serviceDiscovery.PrivateDNSNamespace.ID, "Expected private namespace ID to match")
+		assert.Equal(t, "corp", serviceDiscovery.PrivateDNSNamespace.Name, "Expected private namespace Name to match")
+		assert.Equal(t, "This is a private namespace", serviceDiscovery.PrivateDNSNamespace.Description, "Expected private namespace description to match")
+		assert.Equal(t, "ns-C0VF3F3", serviceDiscovery.PublicDNSNamespace.ID, "Expected public namespace ID to match")
+		assert.Equal(t, "amazon.com", serviceDiscovery.PublicDNSNamespace.Name, "Expected public namespace name to match")
+		assert.Equal(t, "mysds", serviceDiscovery.ServiceDiscoveryService.Name, "Expected SDS Name to match")
```
Comment (SoManyHs) [inline ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: Just for readability, any chance we can extract `serviceDiscovery.ServiceDiscoveryService` into a local var within this block and call the other fields on it?

File: ecs-cli/modules/utils/compose/ecs_params_reader_test.go [hunk 5]
```diff
@@ -512,6 +512,65 @@ task_definition:
...
+
+	err = tmpfile.Close()
+	assert.NoError(t, err, "Could not close tempfile")
+
+	ecsParams, err := ReadECSParams(ecsParamsFileName)
+
+	if assert.NoError(t, err) {
+		serviceDiscovery := ecsParams.RunParams.ServiceDiscovery
+		assert.Equal(t, "nginx", serviceDiscovery.ContainerName, "Expected ContainerName to match")
+		assert.Equal(t, aws.Int64(80), serviceDiscovery.ContainerPort, "Expected ContainerPort to match")
```
Comment (SoManyHs) [inline ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: If we're going to go through the trouble of having an `*int64`, could we perhaps add a test to assert null on this field if not specified? Would also be good to have tests that document expected inputs -- e.g. what makes sense for SRV vs A records.

Comment (allisaurus) [reply ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: I like this idea re: testing inputs; sounds like an opportunity for a table driven test and runs through a bunch of valid (and invalid) scenarios

File: ecs-cli/modules/utils/compose/ecs_params_reader.go [hunk 6]
```diff
@@ -104,6 +105,49 @@ type AwsVpcConfiguration struct {
...
+	ContainerName           string                  `yaml:"container_name"`
+	ContainerPort           *int64                  `yaml:"container_port"`
```
[the remark this answers is not in the document]
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader.go]: Are you suggesting I also switch all of these new string fields to being nil as well? I'm cool with doing that.

File: ecs-cli/modules/utils/compose/ecs_params_reader.go [hunk 7]
```diff
@@ -104,6 +105,49 @@ type AwsVpcConfiguration struct {
...
+	PrivateDNSNamespace     PrivateDNSNamespace     `yaml:"private_dns_namespace"`
+	PublicDNSNamespace      PublicDNSNamespace      `yaml:"public_dns_namespace"`
```
[the remark this answers is not in the document]
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader.go]: They can't be both specified. The plan right now is to include validation logic in `servicediscovery_app.go`. In that file I'll handle the logic of merging the input from flags and from ecs params, and then validating that all of the input makes sense together. 

Flags override ECS Params for the same value, but if you specify a private namespace with one but a public namespace with other- I will quit and log an error.

[the remark this answers is not in the document]
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader.go]: This validation logic will come in a future PR.

File: ecs-cli/modules/utils/compose/ecs_params_reader_test.go (hunk 4 above)
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: Sure

File: ecs-cli/modules/utils/compose/ecs_params_reader_test.go (hunk 5 above)
Comment (PettitWesley) [reply ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: I don't really agree that there's a need to test if the field is nil if its not specified. After all, this test is already a bit superfluous- its basically testing the yaml parsing library the (yaml parsing library has its own tests already). This isn't like the Healthcheck feature which had custom parsing logic. 

> Would also be good to have tests that document expected inputs -- e.g. what makes sense for SRV vs A records.

I agree there should be documentation of that, but I don't think this is the best place for it. 

There will be code that handles what inputs make sense together (pairings of network modes and DNS Records), but that will come in the next PR and it won't be placed in the ECS Params package. The units tests for that will come in `cli/servicediscovery` package.

Comment (allisaurus) [reply ecs-cli/modules/utils/compose/ecs_params_reader_test.go]: I agree it makes more sense to test the SD inputs there; I'm guess that's where the real validation logic will be

File: ecs-cli/modules/utils/compose/ecs_params_reader.go (hunk 1 above)
Comment (SoManyHs) [reply ecs-cli/modules/utils/compose/ecs_params_reader.go]: I don't think there's enough re-use to justify that abstraction, just figured I'd bring it up as food for thought.
